### PR TITLE
Mesa ingestor

### PR DESCRIPTION
**Pull request by jvshields**
It's a mesa model ingester. 🎉 
Should be ready to go. Validated on a hydrogen exhausted solar mesa model and produces a reasonable spectrum as far as I can tell. 

There's some testing infrastructure in here too to give good test coverage that could be applied to the marcs tests at some point in the future.

### CODE REVIEW

File: stardis/base.py [hunk 1]
```diff
@@ -50,12 +51,26 @@ def run_stardis(config_fname, tracing_lambdas_or_nus):
         from stardis.io.model.marcs import read_marcs_model
```
Comment (Knights-Templars) [inline stardis/base.py:51]: Can the imports be made at the top?

File: stardis/base.py [hunk 2]
```diff
@@ -50,12 +51,26 @@ def run_stardis(config_fname, tracing_lambdas_or_nus):
...
         )
         stellar_model = raw_marcs_model.to_stellar_model(
             adata, final_atomic_number=config.model.final_atomic_number
         )
 
+    elif config.model.type == "mesa":
+        from stardis.io.model.mesa import read_mesa_model
+
+        raw_mesa_model = read_mesa_model(Path(config.model.fname))
+        if config.model.truncate_to_shell != -99:
```
Comment (Knights-Templars) [inline stardis/base.py:64]: Why such a significant negative value?

File: stardis/io/model/data/asplund_2009_processed.csv [hunk 3]
```diff
@@ -0,0 +1,70 @@
+,Atom_num,Element,Value
```
Comment (Knights-Templars) [inline stardis/io/model/data/asplund_2009_processed.csv:1]: Stop the index.

Comment (Knights-Templars) [reply stardis/io/model/data/asplund_2009_processed.csv:1]: `atomic_number`

File: stardis/io/model/mesa.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+
+from stardis.model.geometry.radial1d import Radial1DGeometry
+from stardis.model.composition.base import Composition
+
+from stardis.model.base import StellarModel
+from stardis.io.model.util import create_scaled_solar_profile
+
+
+@dataclass
+class MESAModel(object):
```
Comment (Knights-Templars) [inline stardis/io/model/mesa.py:16]: Don't inherit from object.

File: stardis/io/model/mesa.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        )
+
+        atomic_mass_fraction.index.name = "atomic_number"
+        return Composition(density, atomic_mass_fraction)
+
+    def to_stellar_model(
+        self,
+        atom_data,
+        truncate_to_shell_number=None,
+        Y=2.492280e-01,
```
Comment (Knights-Templars) [inline stardis/io/model/mesa.py:86]: Make a constant at the top. Description needed.

File: stardis/io/model/mesa.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+            StellarModel: StellarModel object representing the MESA model.
+        """
+        if truncate_to_shell_number is not None:
+            self.truncate_model(truncate_to_shell_number)
+        mesa_geometry = self.to_geometry()
+        logging.info(
+            f"Creating uniform composition profile from MESA model with Y = {Y} and Z = {Z}."
+        )
+        mesa_composition = self.to_uniform_composition_from_solar(
+            atom_data, Y, Z, final_atomic_number=final_atomic_number
```
Comment (Knights-Templars) [inline stardis/io/model/mesa.py:109]: X --> mass_fraction.

File: stardis/io/model/mesa.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+    Returns:
+        A dictionary containing the metadata extracted from the MESA file.
+
+    Raises:
+        FileNotFoundError: If the specified file does not exist.
+    """
+    # The regular expression pattern
+    METADATA_RE_STR = [
+        ("\s*version_number\s+(\S+)\s*\n", "Version number"),
```
[the remark this answers is not in the document]
Comment (Knights-Templars) [reply stardis/io/model/mesa.py:131]: Remove "\n"

File: stardis/io/model/mesa.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+    metadata_re = [re.compile(re_str[0]) for re_str in METADATA_RE_STR]
+    metadata = {}
+
+    with open(fpath, "rt") as file:
+        contents = file.readlines()
+
+    lines = list(contents)
+
+    for line_number, line in enumerate(lines):
+        match = metadata_re[0].match(line)
```
Comment (Knights-Templars) [inline stardis/io/model/mesa.py:151]:
```suggestion
        match = metadata_re[0].match(line.strip())
```

File: stardis/io/model/mesa.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        if match is not None:
+            STARTING_OF_METADATA = line_number
```
Comment (Knights-Templars) [inline stardis/io/model/mesa.py:153]: Its not a constant. should not be all caps.

File: stardis/io/model/mesa.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+    mesa_shells : int
+        Number of shells to read from the file.
+
+    Returns
+    -------
+    data : pd.DataFrame
+        Data contents of the MESA model file.
+    """
+
+    ROWS_TO_SKIP = 23
```
Comment (Knights-Templars) [inline stardis/io/model/mesa.py:196]: Nice work!

File: stardis/io/model/tests/test_model_io.py [hunk 11]
```diff
@@ -16,6 +21,45 @@ def marcs_model_test_data_file_path():
     return fpath.joinpath(fname)
```
Comment (Knights-Templars) [inline stardis/io/model/tests/test_model_io.py:21]: Fix this with an issue.

File: stardis/io/model/util.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    Args:
+        atom_data: The atom data used to scale the solar mass fractions.
+        Y: The helium abundance. Default is 0.2492280.
+        Z: The metallicity. Default is 0.01337.
+
+    Returns:
+        pandas.DataFrame: The scaled mass fractions.
+
+    """
+    solar_values = pd.read_csv(PATH_TO_ASPLUND_2009)
```
Comment (Knights-Templars) [inline stardis/io/model/util.py:25]: `atomic_number` as index

**Review by Knights-Templars — APPROVED** [review]

File: stardis/io/model/util.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    0.2492280  # The Asplund 2009 mass fraction of measured He
+)
+ASPLUND_DEFAULT_HEAVY_ELEMENTS_MASS_FRACTION_Z = (
+    0.01337  # The Asplund 2009 mass fraction of measured heavy metals
+)
+
+
+def create_scaled_solar_profile(
+    atom_data,
+    Y=ASPLUND_DEFAULT_HELIUM_MASS_FRACTION_Y,
```
Comment (wkerzendorf) [inline stardis/io/model/util.py:16]: i meant rename the variables helium_mass_fraction_y

Comment (jvshields) [reply stardis/io/model/util.py]: It is done

**Review by andrewfullard — APPROVED** [review]
All makes sense to me and has a clean structure

**Pull request merged**